src/GraphQL.Tests/Subscription/SubscriptionExecutionStrategyTests.cs

```diff
@@ -0,0 +1,678 @@
+        Observer.ShouldHaveResult().ShouldBeSimilarTo(@"{ ""data"": { ""testComplex"": {""name"": ""testing"", ""getCounter"": 23 } } }");
+        Observer.ShouldHaveNoMoreResults();
+    }
+
+    [Fact]
+    public async Task RootCancellationThrowsOnInitialExecution()
+    {
+        using var cts = new CancellationTokenSource();
+        cts.Cancel();
+        await Should.ThrowAsync<OperationCanceledException>(async () => await ExecuteAsync("subscription { test }", o => o.CancellationToken = cts.Token));
+    }
+
+    [Fact]
+    public async Task RootCancellationDoesNotAffectSubscriptions()
+    {
+        using var cts = new CancellationTokenSource();
+        var result = await ExecuteAsync("subscription { test }", o => o.CancellationToken = cts.Token);
+        result.ShouldBeSuccessful();
+        Source.Next("hello");
+        Source.Next("testing");
+        cts.Cancel();
+        Source.Next("success");
+        Observer.ShouldHaveResult().ShouldBeSimilarTo(@"{ ""data"": { ""test"": ""hello"" } }");
+        Observer.ShouldHaveResult().ShouldBeSimilarTo(@"{ ""data"": { ""test"": ""testing"" } }");
+        Observer.ShouldHaveResult().ShouldBeSimilarTo(@"{ ""data"": { ""test"": ""success"" } }");
+        Observer.ShouldHaveNoMoreResults();
+    }
+
+    [Fact]
+    public async Task NoEventsAfterSourceDisconnected()
+    {
+        using var cts = new CancellationTokenSource();
+        var result = await ExecuteAsync("subscription { test }", o => o.CancellationToken = cts.Token);
+        result.ShouldBeSuccessful();
+        Source.Next("hello");
+        Source.Next("testing");
+        SubscriptionObj!.Dispose();
+        Source.Next("should not happen");
+        Observer.ShouldHaveResult().ShouldBeSimilarTo(@"{ ""data"": { ""test"": ""hello"" } }");
+        Observer.ShouldHaveResult().ShouldBeSimilarTo(@"{ ""data"": { ""test"": ""testing"" } }");
+        Observer.ShouldHaveNoMoreResults();
+    }
+
+    [Fact]
+    public async Task CheckServiceProvider()
+    {
+        var result = await ExecuteAsync("subscription { testComplex { validateServiceProvider } }", o => o.UserContext["provider"] = o.RequestServices);
+        result.ShouldBeSuccessful();
+        Source.Next("hello");
+        Observer.ShouldHaveResult().ShouldBeSimilarTo(@"{ ""data"": { ""testComplex"": { ""validateServiceProvider"": true } } }");
+    }
+
+    [Fact]
+    public void SubscriptionExecutionStrategy_NullConstructorThrow()
+    {
+        Should.Throw<ArgumentNullException>(() => new SubscriptionExecutionStrategy(null!));
+    }
+
+    [Fact]
+    public async Task CancelWithinErrorHandler_Rethrows()
+    {
+        var result = await ExecuteAsync("subscription { testComplex { nameMayThrowError } }", o =>
+        {
+            o.UnhandledExceptionDelegate = async (context) => throw new InvalidOperationException();
```

Comment:
Happens twice during test run (on different stacks).

Reply:
Yes, I know.  It is very possible that this occurs for other execution strategies also and for the same reason -- throwing an exception within the node's unhandled execution delegate likely bubbles up to the document executer which would handle it again.  i did not verify but it is likely.

Reply:
Nothing bad here I think, just a note.